### PR TITLE
Asset per file scheme, stage 2

### PR DESCRIPTION
**Pull request by Imberflur**
This follows up from https://github.com/Cakefish/bauble/pull/69

### Main changes
- A special `0` identifier is introduced for top level objects so their name does not need to be repeated from the file name.
- Files now require a top level object unless they are empty.
- Local objects can no longer be referenced outside of the current file.
- Local objects no longer registered into `BaubleContext` (similarly to inline objects). E.g. `BaubleContext::assets` now only lists top level objects.
- Object paths now use a use the `ObjectPath` enum instead of being plain `TypePath`s. The enum has 3 variants: `Top`, `Local`, and `Inline`. Uses of a `top_level` bool were replaced with this.

### Other changes
- Resolved various edge cases around registration ordering of assets referencing each other.
- Added tests for these edge cases.
- Fix case where explicit type on binding can be ignored for objects that are a referenced asset that is already registered.
- `path::*::ident` now produces an error when multiple items could be referenced by it instead of selecting one of those items

### CODE REVIEW

File: bauble/src/parse/parser.rs [hunk 1]
```diff
@@ -688,20 +689,49 @@ pub fn parser<'a>() -> impl Parser<'a, ParserSource<'a>, ParseValues, Extra<'a>>
...
                 let is_first = i == 0;
+                let has_top_level_ident = *ident == crate::object_path::TOP_LEVEL_IDENTIFIER;
+
+                let error_emitted = match (is_first, has_top_level_ident) {
+                    (true, true) | (false, false) => false,
+                    (true, false) => {
+                        emitter.emit(Rich::custom(
+                            ident.span,
+                            format!(
+                                "The first item must have '{}' as the identifier",
```
Comment (Imberflur) [inline bauble/src/parse/parser.rs:700]: We now required the first object to be a top level object and subsequent objects must not use the special identifier.

File: bauble/src/parse/value.rs [hunk 2]
```diff
@@ -166,13 +171,43 @@ impl SpannedValue for ParseVal {
...
     pub type_path: Option<Path>,
     pub value: ParseVal,
-    pub is_first: bool,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub enum BindingIdent {
+    /// This is the top level asset in a parsed file.
+    ///
+    /// It has the special cased identifier `0` and appears as the first item in the file..
```
Comment (Imberflur) [inline bauble/src/parse/value.rs]: Extra `.`

File: bauble/src/types/path.rs [hunk 3]
```diff
@@ -222,6 +215,13 @@ fn path_len(path: &str) -> Result<usize> {
         }
     }
 
+    if path.is_empty() {
```
Comment (Imberflur) [inline bauble/src/types/path.rs:218]: moved all the code down past the inner function definitions so it is easier to follow

File: bauble/src/types/path.rs [hunk 4]
```diff
@@ -487,6 +487,7 @@ impl<S: AsRef<str>> TypePath<S> {
             })
     }
 
+    // TODO: can this method be phased out now that `ObjectPath` exists?
```
Comment (Imberflur) [inline bauble/src/types/path.rs]: I need to check if we are still using this.

File: bauble/src/value/convert.rs [hunk 5]
```diff
@@ -223,11 +234,11 @@ pub enum AnyVal<'a> {
...
     pub symbols: &'a Symbols<'a>,
     pub additional_objects: &'a mut AdditionalObjects,
-    // TODO: in stage 2 (of asset per file impl) this probably needs to be `0` rather than anything
-    // that could conflict with the name of a local asset. Because this is used to name inline
-    // objects uniquely.
-    //
     /// Used as a prefix to uniquely name inline objects.
+    ///
+    /// Note, for top level objects we specifically use `0` and not the file name. If we used the
+    /// file name inline object names could collide with inline objects of a local object in
```
Comment (Imberflur) [inline bauble/src/value/convert.rs]:
```suggestion
    /// file name, inline object names could collide with inline objects of a local object in
```

File: bauble/src/value/convert.rs [hunk 6]
```diff
@@ -276,7 +287,7 @@ impl AdditionalObjects {
         &mut self,
         name: TypePathElem<&str>,
         val: Val,
-        symbols: &Symbols,
+        types: &TypeRegistry,
```
Comment (Imberflur) [inline bauble/src/value/convert.rs:290]: Before introducing `EarlySymbols` I worked through understanding all the uses of `Symbols`, this spot and a few others only needed `&TypeRegistry` so changing to that made it easier to get a grasp on what actually needed `Symbols`.

File: bauble/src/value/convert.rs [hunk 7]
```diff
@@ -401,13 +410,14 @@ impl<'a> AdditionalUnspannedObjects<'a> {
...
-    /// appended to the parent object name.
+    /// The closure is passed the file path and the name of the parent object. The generated name
+    /// should be joined to the provided file path. The parent object name can optionally use that
+    /// in its naming logic. Note, a number representing the type id of the current subobject may
+    /// be appended to the parent object name.
     pub fn new_with_custom_namer(
         file_path: TypePath<&'a str>,
         object_name: TypePathElem<&'a str>,
-        namer: &'a mut impl FnMut(TypePathElem<&str>) -> TypePathElem,
+        namer: &'a mut impl FnMut(TypePath<&str>, TypePathElem<&str>) -> ObjectPath,
```
Comment (Imberflur) [inline bauble/src/value/convert.rs:420]: The `namer` closure now has to be responsible for prepending the file path because it also determines the kind of object (top level, local, or inline) so it needs to return a whole `ObjectPath`.

I suppose there are alternative approaches like introducing another enum that contains a `TypePathElem`. But the current approach seems sufficient to me.

File: bauble/src/value/display.rs [hunk 8]
```diff
@@ -253,21 +256,25 @@ fn slice_display<CTX, T: IndentedDisplay<CTX>>(slice: &[T], mut w: LineWriter<CT
 
 impl<CTX: ValueCtx<V>, V: ValueTrait> IndentedDisplay<CTX> for Value<V>
 where
-    V::Ref: std::fmt::Display,
+    V::Ref: std::fmt::Debug,
```
Comment (Imberflur) [inline bauble/src/value/display.rs:259]: I removed the `Display` requirement because it would be a bit hacky for `ObjectPath` to have a `Display` impl dedicated for the purposes here. Instead, the display logic is moved to the `ValueCtx` trait. Replacing `ValueCtx::inline` with `ValueCtx::display_ref`.

`Debug` is still needed for diagnostics when errors occur.

File: bauble/src/value/display.rs [hunk 9]
```diff
@@ -253,21 +256,25 @@ fn slice_display<CTX, T: IndentedDisplay<CTX>>(slice: &[T], mut w: LineWriter<CT
...
+                Some(DisplayRef::Path(path)) => {
                     w.write("$");
-                    w.fmt(r);
+                    w.fmt(path);
                 }
-            }
+                None => {
+                    // TODO: this is an error (empty path or missing inline object), but
+                    // IndentedDisplay has no path for reporting errors.
+                    w.write("ERROR {r:?} is empty or points to missing inline object");
```
Comment (Imberflur) [inline bauble/src/value/display.rs:275]: Previously this just wrote out the path if an inline object didn't exist, now we write out an error message into the output since the current setup has no other allowance for reporting errors.

Maybe a field could be added to the `LineWriter` struct to collect any errors.

File: bauble/src/value/early_context.rs [hunk 10]
```diff
@@ -0,0 +1,374 @@
+use crate::context::{BaubleContext, PathReference};
```
Comment (Imberflur) [inline bauble/src/value/early_context.rs:1]: This whole module was added because we had an issue where a reference written in a file that was based on `use`s in that file or based on referencing locally defined items could not be resolved properly to a full path until the referenced asset was registered.

Registering an asset required knowing the type of the asset, which might require other assets to already be registered. In this case we can delay registering the asset, however when we handle delayed assets `Symbols` is not available, so any paths must already be fully resolved. This created a conundrum.

So we introduce a earlier pass where assets are registered without their types. And we use the context built there to fully resolve paths where needed in the next pass where we resolve the types of assets.

The types here are copies of the ones used in `BaubleContext` adjusted for this purpose. Simplifications are made where possible because this needs to support less functionality than `BaubleContext`. This `EarlyContext` is ephemeral, it is created and discarded during loading of new/updated bauble files.

Note, an alternative approach is making the type field for assets in `BaubleContext` optional. This was found to be less appealing. We expect that the field will always be `Some` after loading is complete, and it is more robust to have this encoded in the type system.

File: bauble/src/value/error.rs [hunk 11]
```diff
@@ -11,18 +10,26 @@ use crate::{
...
+    pub(super) module: bool,
+}
+
 #[derive(Clone, Debug)]
 pub struct RefError {
-    pub(super) uses: Option<HashMap<TypePathElem, PathReference>>,
+    pub(super) uses: Option<HashSet<TypePathElem>>,
     pub(super) path: PathKind,
-    pub(super) path_ref: PathReference,
+    pub(super) path_ref: Option<ErrorPathReference>,
```
Comment (Imberflur) [inline bauble/src/value/error.rs:24]: We have a bunch of different `PathReference` like types now, so I introduced one dedicated for errors that has the common information that the errors actually utilize. Also many places creating a `RefError` just created a default value, so I switched this to an `Option` to make that simpler and more explicit.

File: bauble/src/value/mod.rs [hunk 12]
```diff
@@ -1,28 +1,31 @@
+mod convert;
+mod display;
+mod early_context;
+mod error;
+mod symbols;
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:5]: I'm trying to slowly adjust our code to have modules at the top.

File: bauble/src/value/mod.rs [hunk 13]
```diff
@@ -555,50 +553,40 @@ impl std::fmt::Display for PathKind {
...
         let mut errors = Vec::new();
         let old_len = delayed.len();
 
         // Try to register delayed registers, and remove them as they succeed.
         delayed.retain(|d| {
-            if let Some(r) = match &d.reference.value {
-                PathKind::Direct(path) => ctx.get_ref(path.borrow()),
-                // TODO: what if indirect path becomes ambiguous due to later registered items that
-                // were delayed?
-                PathKind::Indirect(path, ident) => {
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:597]: Iirc I resolved this TODO by having the full path resolved beforehand (which iirc is enabled by the extra pass to register objects without knowing their types)

File: bauble/src/value/mod.rs [hunk 14]
```diff
@@ -615,7 +603,13 @@ pub(crate) fn resolve_delayed(
                     let Some(desired_ty) = (match path {
                         PathKind::Direct(path) => ctx.get_ref(path.borrow()),
                         PathKind::Indirect(path, ident) => {
-                            ctx.ref_with_ident(path.borrow(), ident.borrow())
+                            match ctx.ref_with_ident(path.borrow(), ident.borrow()) {
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:606]: `ref_with_ident` now produces an error if the path is ambiguous, so we need to handle that.

File: bauble/src/value/mod.rs [hunk 15]
```diff
@@ -669,55 +669,52 @@ pub(crate) fn resolve_delayed(
...
-                            .spanned(scc[0].span),
+                            .spanned(referer.span),
                         )
                     } else {
-                        // TODO: Is this path possible? Wouldn't Ref have to refer to itself for
-                        // scc.len() == 1?
+                        // We make sure that the referenced asset exists in the pre-registered
+                        // assets before producing `DelayedRegister`. So this error will only occur
+                        // if the referenced asset's type failed to resolve (such that it wasn't
+                        // registered in resolve_delayed).
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:699]: After writing the removed TODO comment, I looked more into how `tarjan_scc` works, and learned that a single node that doesn't participate in any cycles is considered strongly connected component by itself.

File: bauble/src/value/mod.rs [hunk 16]
```diff
@@ -726,82 +723,120 @@ pub(crate) fn resolve_delayed(
...
-/// Normally, the path is just the files's bauble path joined with the object name.
+/// The top level object in each file will have a path that matches the path of the file containing
+/// it. The identifier for these objects is always "0" and it can be referred to locally in the
+/// same file using this identifier.
 ///
-/// If an object is the first in the file and its name matches the file name, it receives a special
-/// path that is just the file's bauble path.
-fn object_path(
+/// TODO: make sure this is updated
+/// For other objects, the path is just the files's bauble path joined with the object identifier.
```
Comment (Imberflur) [inline bauble/src/value/mod.rs]: I think this `TODO` can be removed, maybe this can mention the `Top` and `Local` variants of `ObjectPath`.

File: bauble/src/value/mod.rs [hunk 17]
```diff
@@ -726,82 +723,120 @@ pub(crate) fn resolve_delayed(
...
+    (ident, path)
+}
+
+/// Registers all new top level asset paths into [`EarlyContext`] so they will be known for
+/// resolving full paths from `use`s in [`register_assets`].
+///
+/// We need to know what items brought into scope with `use` are assets rather than types or
+/// modules to properly dertermine the full path (otherwise there could be multiple candidates
+/// because items from different namespaces can have the same name).
+pub(crate) fn pre_register_assets(
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:752]: This is the new pass to register objects without their types being known.

File: bauble/src/value/mod.rs [hunk 18]
```diff
@@ -726,82 +723,120 @@ pub(crate) fn resolve_delayed(
...
     for use_path in &values.uses {
-        // TODO: This will error if the referenced file is yet to be passed register_assets
         if let Err(e) = symbols.add_use(use_path) {
             errors.push(e);
         }
     }
 
-    // Move `uses` in/out of `Symbols` every loop so we have mutable access to `ctx` at certain
-    // points.
-    let Symbols { mut uses, .. } = symbols;
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:772]: I made `EarlySymbols` hold a mutable `ctx` reference so we can avoid this and simplify the code here.

File: bauble/src/value/mod.rs [hunk 19]
```diff
@@ -726,82 +723,120 @@ pub(crate) fn resolve_delayed(
...
+            // registered when registering an object of type `T`. So the when
+            // the referenced object has yet to be registered, the reference
+            // type may not exist. Thus, trying to resolve the type at this
+            // point can fail.
+            && !matches!(&*binding.value.value, Value::Ref(_))
         {
             symbols.resolve_type(ty)
         } else {
-            let res = value_type(&binding.value, &symbols)
+            let res = convert::value_type(&binding.value, &symbols)
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:836]: This is a small stylistic preference to `use` the module a method is in rather than the method directly to help readability.

File: bauble/src/value/mod.rs [hunk 20]
```diff
@@ -811,19 +846,33 @@ pub(crate) fn register_assets(
...
+                && let Value::Ref(ref_path) = &*binding.value.value
+                // Note: If this fails, then `value_type()` would have produced the same error
+                // already in `res` from calling `symbols.resolve_asset_type()` which uses
+                // `resolve_path` internally. So there is no extra information from this error.
+                //
+                // We use `resolve_asset` instead of just `resolve_path` because the asset should
+                // exist due to `pre_register_assets` or we will invevitably produce an error
+                // anyway, and the errors produced in register_assets are better than
+                // `resolve_delayed` because `symbols.uses` is available.
+                && let Ok((maybe_ty, reference)) = symbols.resolve_asset(ref_path)
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:866]: Here I justify why we can ignore the error message.

File: bauble/src/value/mod.rs [hunk 21]
```diff
@@ -835,42 +884,40 @@ pub(crate) fn register_assets(
...
-                    },
-                ) {
-                    errors.push(e.spanned(binding.value.value.span));
-                }
-                Symbols { uses, .. } = symbols;
+            if let Ok(_res) = res
+                // We skipped `resolve_type` above because `Ref<T>` might not be registered, but if we
+                // found referenced asset in `value_type`, then the asset is already registered so
+                // `Ref<T>` will either also be registered or it will be the wrong type.
+                && let Some(ty) = &binding.type_path
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:899]: We need to fall back to the logic skipped by `&& !matches!(&*binding.value.value, Value::Ref(_))` when the binding has an explicit type and we didn't take the `DelayedRegister` code path. Otherwise, we can end up not checking that the type of the referenced asset matches the explicitly written type.

File: bauble/src/value/mod.rs [hunk 22]
```diff
@@ -883,129 +930,115 @@ pub(crate) fn register_assets(
 pub(crate) fn convert_values(
     file: FileId,
     values: ParseValues,
-    default_symbols: &Symbols,
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:886]: `default_symbols` was unused, so I removed it

I believe this was a legacy mechanism that was replaced with registering redirects in the root node of `BaubleContext`.

File: bauble/src/value/mod.rs [hunk 23]
```diff
@@ -883,129 +930,115 @@ pub(crate) fn register_assets(
...
-            use_errors.push(ConversionError::UnregisteredAsset.spanned(span));
-        }
-    }
+        let asset = match path {
+            ObjectPath::Top(path) => symbols.ctx.get_ref(path.borrow()).and_then(|r| r.asset),
+            ObjectPath::Local(path) => local_ctx.get(path.borrow()).map(|ty| (ty, path.clone())),
+            ObjectPath::Inline(_) => unreachable!("can't be returned from object_ident_path"),
+        };
 
-    symbols.add(use_symbols);
```
Comment (Imberflur) [inline bauble/src/value/mod.rs:926]: `use`s are now added to symbols before adding local assets, and conflicts now produce errors instead of ignoring them (`Symbols::add` would simply override anything that conflicted)

File: bauble/src/value/symbols.rs [hunk 24]
```diff
@@ -3,23 +3,167 @@ use std::{borrow::Cow, collections::HashMap};
...
+        "Cannot use path::*::ident syntax in generic parameters",
+    ))
+    .spanned(span)
+}
+
+/// `PathReference` for `Symbols` `uses`. Distinguishes between local and top level objects.
+#[derive(Default, Clone)]
+struct UseReference {
+    ty: Option<TypeId>,
+    asset: Option<(TypeId, ObjectPath)>,
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs:52]: `PathReference` is only used for top level objects so it just uses `TypePath`. So here we now have a dedicated type for internal usage in `Symbols` instead of reuseing `PathReference`.

File: bauble/src/value/symbols.rs [hunk 25]
```diff
@@ -3,23 +3,167 @@ use std::{borrow::Cow, collections::HashMap};
...
+            self.module = other.module;
+        }
+    }
+}
+
+/// `CombinedPathReference` for `EarlySymbols` `uses`. Distinguishes between local and top level objects.
+#[derive(Default, Clone, Debug)]
+struct EarlyUseReference {
+    ty: Option<TypeId>,
+    asset: Option<(Option<TypeId>, ObjectPath)>,
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs:85]: Like `UseReference` except the type becomes optional. A common pattern in this PR is introducing variants on the `PathReference` type for different purposes:
- `EarlyPathReference` (used internally in `EarlyContext` where the type of an object is definitely not known)
- `CombinedPathReference` (returned from queries on `EarlyContext`, where the type is optional and will be known if the object is registered in the inner `BaubleContext`)
- `UseReference` (internal use in `Symbols`, can reference local or top level objects)
- `EarlyUseReference` (internal use in `EarlySymbols`, object type is optional)
- `ErrorPathReference` (common information about items referenced by a path for errors)

File: bauble/src/value/symbols.rs [hunk 26]
```diff
@@ -3,23 +3,167 @@ use std::{borrow::Cow, collections::HashMap};
...
+            module: reference.module.is_some(),
+        }
+    }
+}
 
 /// Representation of item names available in the current module.
 ///
 /// There are multiple namespaces: types, assets (i.e. values defined in bauble), and modules.
-#[derive(Clone)]
+//#[derive(Clone)]
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs]: this derive can be removed completely

File: bauble/src/value/symbols.rs [hunk 27]
```diff
@@ -30,24 +174,42 @@ impl<'a> Symbols<'a> {
...
 
         Ok(())
     }
 
-    pub fn add(&mut self, symbols: Symbols) {
-        // TODO: what about conflicting entries?
-        self.uses.extend(symbols.uses)
+    /// Note, this can also add the top level object of the current file under its local
+    /// identifier.
+    pub fn add_local_object(
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs]: `add_ref` is now private and instead we expose this more specific method.

Comment (Imberflur) [reply bauble/src/value/symbols.rs]: I might need to rename this, I forgot it would be used for the top level object's local identifier when first implementing it.

File: bauble/src/value/symbols.rs [hunk 28]
```diff
@@ -207,156 +373,645 @@ impl<'a> Symbols<'a> {
...
-                    .map_err(|p| p.spanned(raw_path.span()))?;
-                PathKind::Direct(leading)
+                if !matches!(kind, ResolveKind::Type) {
+                    return Err(generic_with_non_type_error(raw_path));
+                }
+
+                // The outer type and inner parameter in generic types need to be fully expanded
+                // based on `uses` here (since the resolved path for a concrete instance of the
+                // generic type only makes sense to lookup via full paths in the context and not
+                // via `uses`).
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs:407]: I removed the inner `resolve_path` defined in `resolve_item` below and I think also fixed https://github.com/Cakefish/bauble/issues/57.

Now when encountering a path with generics, the main `resolve_path` method will eagerly resolve the outer and inner paths based on `uses` rather than waiting for this to happen in `resolve_item`.

The reason we can't do this for everything is that the first element of a path can reference both things from `uses` and from the root of the `BaubleContext`. So if we resolve to a full path based on `uses`, then in `resolve_item` we will miss anything that the original path referenced in the root of `BaubleContext`. This is fine if we consider a single namespace since `uses` takes priority. However, for errors we want to know the item in each namespace that a path could have referenced (iirc this is part of the reasoning, the logic here is a bit complex...). Additionally, delaying the lookup in `uses` is useful because this returns `PathKind` which would need to be replaced with something that also distinguishes top level vs local objects.

For generics we know we are always referencing a type.

File: bauble/src/value/symbols.rs [hunk 29]
```diff
@@ -207,156 +373,645 @@ impl<'a> Symbols<'a> {
...
-                .unwrap_or((raw_path, &raw_path));
-
-            let path = symbols.resolve_path(path)?;
-
-            Ok(if matches!(ref_kind, RefKind::Type) {
-                match path.value {
-                    PathKind::Direct(type_path) => {
-                        if let Some(r) = symbols.uses.get(type_path.as_str())
+        kind: ResolveKind,
+    ) -> Result<(Cow<'_, UseReference>, PathKind)> {
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs:452]: `resolve_item` now returns the resolved `PathKind` instead of the caller re-resolving for errors, this reduces the places calling `resolve_path` which makes the logic easier to follow.

File: bauble/src/value/symbols.rs [hunk 30]
```diff
@@ -207,156 +373,645 @@ impl<'a> Symbols<'a> {
...
+        }
+    }
+}
+
+/// Representation of item names available in the current module.
+///
+/// This is used before types of assets are fully resolved. Afterwards, [`Symbols`] can be used.
+///
+/// There are multiple namespaces: types, assets (i.e. values defined in bauble), and modules.
+pub(crate) struct EarlySymbols<'a, 'b> {
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs:546]: `EarlySymbols` is basically the `Symbols` equivalent of `EarlyContext`, most of the code below is just duplicated logic from `Symbols` except with object types being optional

File: bauble/src/value/symbols.rs [hunk 31]
```diff
@@ -207,156 +373,645 @@ impl<'a> Symbols<'a> {
...
+                path: resolved_path,
+                path_ref: Some(item.into_owned().into()),
                 kind: RefKind::Type,
             }))
             .spanned(path.span()))
         }
     }
 }
+
+pub(crate) trait SymbolsCommon {
```
Comment (Imberflur) [inline bauble/src/value/symbols.rs:975]: A few methods are used both where we have `Symbols` and where we have an `EarlySymbols`. This trait abstracts over the necessary functionality that those methods need from the symbols type.

File: bauble/src/context.rs [hunk 32]
```diff
@@ -11,27 +12,13 @@ pub type Source = ariadne::Source<String>;
...
-}
-
 /// A type containing multiple references generally derived from a path.
 #[derive(Default, Clone, Debug)]
 pub struct PathReference {
     /// The type referenced by a path.
     pub ty: Option<TypeId>,
     /// The asset (and its path) referenced by the path.
-    pub asset: Option<(TypeId, TypePath, AssetKind)>,
+    pub asset: Option<(TypeId, TypePath)>,
```
Comment (Imberflur) [inline bauble/src/context.rs:21]: No kind is needed and this still just uses `TypePath` because it now only refers to top level objects.

File: bauble/src/context.rs [hunk 33]
```diff
@@ -73,23 +60,17 @@ impl PathReference {
         })
     }
 
-    /// Overrides references of `self` with references of `other`, returns true if
-    /// anything was overriden.
-    pub fn combine_override(&mut self, other: Self) -> bool {
-        let mut o = false;
+    /// Overrides references of `self` with references of `other`.
+    pub fn combine_override(&mut self, other: Self) {
```
Comment (Imberflur) [inline bauble/src/context.rs:64]: Return value was never used

File: bauble/src/context.rs [hunk 34]
```diff
@@ -418,11 +390,9 @@ impl CtxNode {
...
         });
 
-        // Top level assets match the path of their file so this will be from this file if it is
-        // top level.
-        self.reference
-            .asset
-            .take_if(|(_, kind)| matches!(kind, AssetKind::TopLevel));
+        // Top level assets match the path of their file (and all assets registered in
+        // BaubleContext are top level).
+        self.reference.asset.take();
```
Comment (Imberflur) [inline bauble/src/context.rs:395]: `clear_file_assets` basically just needs to only clear the top level object now

File: bauble/src/context.rs [hunk 35]
```diff
@@ -526,36 +496,27 @@ impl BaubleContext {
         self.files.iter().map(|e| (e.0.borrow(), e.1.text()))
     }
 
+    /// Given an asset type registers the internal Ref type for that asset and returns its ID.
+    pub(crate) fn register_asset_ref_ty(&mut self, ty: TypeId) -> TypeId {
```
Comment (Imberflur) [inline bauble/src/context.rs:500]: Extracted to a new method for reuse in `LocalContext` iirc

File: bauble/src/context.rs [hunk 36]
```diff
@@ -642,15 +603,34 @@ impl BaubleContext {
             }
         }
 
+        let mut early_ctx = crate::value::EarlyContext::new(self);
```
Comment (Imberflur) [inline bauble/src/context.rs:606]: Here and below are the main additions/changes to the logic when loading files.

File: bauble/src/context.rs [hunk 37]
```diff
@@ -728,32 +708,46 @@ impl BaubleContext {
     /// `ident`.
     ///
     /// Returns the [`PathReference`] from [`CtxNode::reference`]. If multiple nodes are found with
-    /// `ident`, the refs from these will be combined (potentially overriding each other).
-    //
-    // TODO: couldn't this overriding lead to unexpected behavior? Should we return an error when
-    // there are multiple results in the same namespace?
```
Comment (Imberflur) [inline bauble/src/context.rs:734]: TODO resolved by producing an error on ambiguity

File: bauble/src/local_context.rs [hunk 38]
```diff
@@ -0,0 +1,42 @@
+use crate::context::BaubleContext;
+use crate::path::TypePath;
+use crate::types::TypeId;
+use indexmap::IndexMap;
+
+/// Transient context that is built and used when loading file(s).
+///
+/// Contains registery of local objects. These are objects that are only referencable by other
+/// objects in the same file.
+pub(crate) struct LocalContext {
```
Comment (Imberflur) [inline bauble/src/local_context.rs:10]: Local objects are now registered in this ephemeral context type instead of `BaubleContext`. It only exists during the loading process.

This ended up being a super simple type!

File: bauble/src/object_path.rs [hunk 39]
```diff
@@ -0,0 +1,181 @@
...
+
+/// Special cased identifier that is required and only allowed for the first asset in a file
+/// (i.e. the top level asset that is named after the file).
+pub const TOP_LEVEL_IDENTIFIER: &str = "0";
+
+/// Full path to a bauble object (or external asset).
+///
+/// Path format documented in [`TypePath`].
+#[derive(Copy, Clone, Debug)]
+pub enum ObjectPath<S = String> {
```
Comment (Imberflur) [inline bauble/src/object_path.rs:41]: New `ObjectPath` enum, I hope the docs are sufficient explanation.

File: bauble/src/traits.rs [hunk 40]
```diff
@@ -243,8 +244,8 @@ pub trait BaubleAllocator<'a> {
     where
         Self: 'a;
 
-    /// The inner representation of a type path created by this allocator.
-    type TypePathInner: 'static;
+    /// The inner representation of an object path created by this allocator.
+    type PathInner: 'static;
```
Comment (Imberflur) [inline bauble/src/traits.rs:248]: renamed for clarity

File: bauble/tests/integration.rs [hunk 41]
```diff
@@ -154,23 +155,23 @@ fn test_load_partial(
 fn new_nested_reload_paths() {
     let a = &test_file!(
         "a",
-        r#"test = integration::Test { x: -5, y: 5 }"#,
+        r#"0 = integration::Test { x: -5, y: 5 }"#,
```
Comment (Imberflur) [inline bauble/tests/integration.rs:158]: all the tests needed updating for the new syntax

**Review by Imberflur — COMMENTED** [review]
Here are notes for reviewers, and a few cleanup tasks for myself.

File: bauble/src/object_path.rs (hunk 39 above)
Comment (MavethGH) [reply bauble/src/object_path.rs:41]: It might be helpful to have examples of what the contents of each variant might look like.

**Comment by MavethGH** [comment]
Testing whether I can comment without it being on a specific line.

**Review by MavethGH — APPROVED** [review]

File: bauble/src/object_path.rs (hunk 39 above)
Comment (Imberflur) [reply bauble/src/object_path.rs:41]: Added a set of examples

File: bauble/src/value/symbols.rs (hunk 27 above)
Comment (Imberflur) [reply bauble/src/value/symbols.rs]: Renamed to `add_current_file_object`

File: bauble/src/value/symbols.rs (hunk 26 above)
Comment (Imberflur) [reply bauble/src/value/symbols.rs]: removed

File: bauble/src/value/mod.rs (hunk 16 above)
Comment (Imberflur) [reply bauble/src/value/mod.rs]: removed and rewrote notes a bit

File: bauble/src/types/path.rs (hunk 4 above)
Comment (Imberflur) [reply bauble/src/types/path.rs]: removed method

File: bauble/src/parse/value.rs (hunk 2 above)
Comment (Imberflur) [reply bauble/src/parse/value.rs]: fixed
